flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/utils/AsyncUtil.java

```diff
@@ -0,0 +1,216 @@
+
+        public final int asyncBufferCapacity;
+        public final long asyncTimeout;
+        public final AsyncDataStream.OutputMode asyncOutputMode;
+        public final AsyncRetryStrategy<RowData> asyncRetryStrategy;
+
+        @JsonCreator
+        public Options(
+                int asyncBufferCapacity,
+                long asyncTimeout,
+                AsyncDataStream.OutputMode asyncOutputMode,
+                AsyncRetryStrategy<RowData> asyncRetryStrategy) {
+            this.asyncBufferCapacity = asyncBufferCapacity;
+            this.asyncTimeout = asyncTimeout;
+            this.asyncOutputMode = asyncOutputMode;
+            this.asyncRetryStrategy = asyncRetryStrategy;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            Options that = (Options) o;
+            return asyncBufferCapacity == that.asyncBufferCapacity
+                    && asyncTimeout == that.asyncTimeout
+                    && asyncOutputMode == that.asyncOutputMode;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(asyncBufferCapacity, asyncTimeout, asyncOutputMode);
+        }
+
+        @Override
+        public String toString() {
+            return asyncOutputMode + ", " + asyncTimeout + "ms, " + asyncBufferCapacity;
+        }
+    }
+
+    private static AsyncRetryStrategy<RowData> getResultRetryStrategy(
+            ExecutionConfigOptions.RetryStrategy retryStrategy,
+            Duration retryDelay,
+            int retryMaxAttempts) {
+        // Only fixed delay is allowed at the moment, so just ignore the config.
+        return new AsyncRetryStrategies.FixedDelayRetryStrategyBuilder<RowData>(
+                        retryMaxAttempts, retryDelay.toMillis())
+                .ifResult(EMPTY_RESPONSE)
+                .ifException(ANY_EXCEPTION)
+                .build();
+        //        return NO_RETRY_STRATEGY;
+    }
+
+    private static class FunctionFinder extends RexDefaultVisitor<Boolean> {
+
+        private final boolean findAsyncCall;
+        private final boolean recursive;
+
+        public FunctionFinder(boolean findAsyncCall, boolean recursive) {
+            this.findAsyncCall = findAsyncCall;
+            this.recursive = recursive;
+        }
+
+        @Override
+        public Boolean visitNode(RexNode rexNode) {
+            return false;
+        }
+
+        private boolean isImmediateAsyncCall(RexCall call) {
+            if (!(call.getOperator() instanceof BridgingSqlFunction)) {
+                return false;
+            }
+            FunctionDefinition definition =
+                    ((BridgingSqlFunction) call.getOperator()).getDefinition();
+            return AsyncScalarFunction.class.isAssignableFrom(definition.getClass());
```

Comment:
Use `ShortcutUtils#unwrapFunctionDefinition(org.apache.calcite.rex.RexNode)` instead. The check should be based purely on the `FunctionDefinition.getKind()`. This allows to work with FunctionDefinition and defer the need for the actual class in the classpath during planning. See also `SpecializedFunction`.

Reply:
Exactly, very useful when you want to build services around it ;-)